### ISSUE TITLE
Rules are not redefined on hot load

### ISSUE THREAD

**Issue by kennyjwilli** [CoNarrative/precept#80]
Currently in order to get rules to be redefined you must refresh your entire page. Development would be much easier if a "reloaded" workflow was supported. 

I have just tested this with `precept.rules/rules`. It's probable that other macros are affected as well (`defsub`, `define`, etc.)

**Comment by levand** [comment]
If I can get #75 working, this should be easy (since you can just throw away the session and start working with a new one.)

**Comment by alex-dixon** [comment]
Related: state can get blown away on figwheel refresh. #75 can provide options for the user to manage. In the meantime (or if we continue to store state at the implementation level) we might experiment with using `defonce` instead of `def` for atoms that are internal.

**Comment by alex-dixon** [comment]
Brought this up in [Clara #316](https://github.com/cerner/clara-rules/issues/316) because it may be related to the issue there (deleting rules while developing requires figwheel to be restarted).

**Comment by alex-dixon** [comment]
The following is CLJS specific:

It appears we've made significant progress. When we remove or rename a rule, it's removed from the session when figwheel reloads. This means we can comment out/rename/delete a rule, save the file, and see the resulting change in our browser without having to refresh the page. When we uncomment the same rule and save the file, we can see it has been added to the session. Again, no need to refresh the page, just save the file and figwheel will detect the change and reload.

At this point, I want to figure out what we want the global state behavior to be when adding or removing a rule. At present, facts in the session are wiped whenever a change to a rule is made. I imagine we'd like the state to remain, but for the resulting state to include the effects of any changes to the rules.

We'll need to make the following changes:
1. Index one-to-many-facts
2. Index all facts according to the way in which they were inserted (logically vs unconditionally)
3. Read all facts from the index that were inserted unconditionally

The procedure for reloading should be able to be handled in a single function that sits below the session definition in the rule file. Not entirely confident about the order here but should be something like:
1. Clear the env/*compiler* of all productions
2. Insert unconditionally all facts taken from the index (unconditional inserts only).
3. Clear the `fact-index`
4. Fire rules
